File: app/src/main/java/com/chopcode/rutago/app/managers/core/ratings/RatingManager.java
package com.chopcode.rutago.app.managers.core.ratings;

import android.util.Log;
import com.chopcode.rutago.app.config.MyApp;
import com.chopcode.rutago.app.models.Rating;
import com.chopcode.rutago.app.models.Reservation;
import com.google.firebase.database.DatabaseReference;
import java.util.HashMap;
import java.util.Map;

/**
 * ⭐ Rating Manager (Singleton)
 *
 * Responsable de la gestión del feedback y reputación de los conductores.
 * Responsabilidades:
 * - Orquestar la creación de calificaciones vinculadas a una reserva específica.
 * - Asegurar la consistencia dual: guarda la reseña en el nodo de calificaciones y 
 *   marca la reserva como "calificada" simultáneamente.
 * - Implementar lógica de persistencia bilingüe (Mapeo dual para Firebase).
 */
public class RatingManager {
    private static final String TAG = "RatingManager";
    private static RatingManager instance;
    private final DatabaseReference ratingsRef;
    private final DatabaseReference reservesRef;

    private RatingManager() {
        this.ratingsRef = MyApp.getDatabaseReference("calificaciones_conductores");
        this.reservesRef = MyApp.getDatabaseReference("reservas");
    }

    /**
     * @return Instancia única del manager.
     */
    public static synchronized RatingManager getInstance() {
        if (instance == null) instance = new RatingManager();
        return instance;
    }

    public interface RatingCallback {
        void onSuccess();
        void onError(String error);
    }

    /**
     * Registra una nueva calificación para un viaje finalizado.
     * @param reservation Objeto reserva que sirve como contexto.
     * @param stars Puntuación numérica otorgada (1-5).
     * @param comment Observaciones cualitativas opcionales.
     */
    public void calificarViaje(Reservation reservation, float stars, String comment, RatingCallback callback) {
        if (reservation == null || reservation.getDriverId() == null) {
            if (callback != null) callback.onError("Datos de reserva inválidos para calificar.");
            return;
        }

        String driverId = reservation.getDriverId();
        String userId = MyApp.getCurrentUserId();

        Rating rating = new Rating(
                userId,
                reservation.getName(),
                driverId,
                reservation.getIdReservation(),
                reservation.getRouteName(),
                stars,
                comment
        );

        DatabaseReference ref = ratingsRef.child(driverId).push();
        rating.setId(ref.getKey());

        // 1. Guardar la calificación en el nodo maestro de reputación
        ref.setValue(rating)
                .addOnSuccessListener(aVoid -> markAsRated(reservation.getIdReservation(), stars, callback))
                .addOnFailureListener(e -> { if (callback != null) callback.onError(e.getMessage()); });
    }

    /**
     * Actualiza la reserva original para reflejar que ya ha sido calificada, 
     * evitando duplicidad de feedback.
     */
    private void markAsRated(String reservationId, float stars, RatingCallback callback) {
        if (reservationId == null) return;
        
        Map<String, Object> updates = new HashMap<>();
        updates.put("rated", true);
        updates.put("calificada", true);
        updates.put("rating", stars);
        updates.put("calificacion", stars);

        reservesRef.child(reservationId).updateChildren(updates)
                .addOnSuccessListener(aVoid -> { if (callback != null) callback.onSuccess(); })
                .addOnFailureListener(e -> { 
                    Log.e(TAG, "❌ Error al marcar reserva como calificada: " + e.getMessage());
                    if (callback != null) callback.onError("No se pudo actualizar el estado de la reserva.");
                });
    }
}

File: app/src/main/java/com/chopcode/rutago/app/managers/core/ratings/RatingManager.kt
package com.chopcode.rutago.app.managers.core.ratings

import android.util.Log
import com.chopcode.rutago.app.config.MyApp
import com.chopcode.rutago.app.models.Rating
import com.chopcode.rutago.app.models.Reservation
import com.google.firebase.database.DatabaseReference

/**
 * ⭐ Rating Manager (Singleton)
 * Responsable de la gestión del feedback y reputación de los conductores.
 * Normalizado a Inglés con soporte legacy.
 */
class RatingManager private constructor() {

    private val ratingsRef: DatabaseReference = MyApp.getDatabaseReference("calificaciones_conductores")
    private val reservesRef: DatabaseReference = MyApp.getDatabaseReference("reservas")

    companion object {
        @Volatile
        private var instance: RatingManager? = null

        fun getInstance(): RatingManager {
            return instance ?: synchronized(this) {
                instance ?: RatingManager().also { instance = it }
            }
        }
    }

    interface RatingCallback {
        fun onSuccess()
        fun onError(error: String?)
    }

    /**
     * Registra una nueva calificación para un viaje finalizado.
     */
    fun calificarViaje(reservation: Reservation, stars: Float, comment: String, callback: RatingCallback?) {
        val driverId = reservation.driverId.ifEmpty { return }
        val userId = MyApp.getCurrentUserId() ?: return

        val rating = Rating(
            userId,
            reservation.passengerName,
            driverId,
            reservation.id,
            reservation.origin + " → " + reservation.destination,
            stars,
            comment
        )

        val ref = ratingsRef.child(driverId).push()
        rating.id = ref.key

        ref.setValue(rating).addOnSuccessListener {
            markAsRated(reservation.id, stars, callback)
        }.addOnFailureListener { e ->
            callback?.onError(e.message)
        }
    }

    private fun markAsRated(reservationId: String, stars: Float, callback: RatingCallback?) {
        val updates = mapOf(
            "isRated" to true,
            "rated" to true, // Legacy
            "calificada" to true, // Legacy
            "rating" to stars,
            "calificacion" to stars // Legacy
        )

        reservesRef.child(reservationId).updateChildren(updates)
            .addOnSuccessListener { callback?.onSuccess() }
            .addOnFailureListener { e ->
                Log.e("RatingManager", "❌ Error al marcar reserva: ${e.message}")
                callback?.onError("No se pudo actualizar el estado de la reserva.")
            }
    }
}
